bu0836.1
'\"
'\" bu0836(1) man page
'\"
'\" Copyright (C) 2010  Melchior FRANZ  <melchior.franz@gmail.com>
'\"
'\" This program is free software; you can redistribute it and/or
'\" modify it under the terms of the GNU General Public License as
'\" published by the Free Software Foundation; either version 2 of the
'\" License, or (at your option) any later version.
'\"
'\" This program is distributed in the hope that it will be useful, but
'\" WITHOUT ANY WARRANTY; without even the implied warranty of
'\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
'\" General Public License for more details.
'\"
'\" You should have received a copy of the GNU General Public License
'\" along with this program; if not, write to the Free Software
'\" Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
'\"
.MA ""
.TH BU0836 1 "30 April 2010" "v0.2"
'\"
'\" left align & no hyphenation
.ad l
.nh
'\"
'\"
'\"""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
.SH NAME
bu0836 \- configuration utility for joystick controllers of the BU0836 family
'\"
'\"
'\"
'\"""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
.SH SYNOPSIS
\fBbu0836\fR [\fB\-v\fR] [\fB\-d\fI device\fR] [\fB\-l\fR] [\fB\-s\fR] [\fB\-m\fR]
.br
\fBbu0836\fR [\fB\-v\fR] [\fB\-d\fI device\fR] [\fB\-r\fR]
[\fB\-a\fI list\fR [\fB\-i \fIinvert\fR] [\fB\-z \fIzoom\fR]]
[\fB\-b\fI list\fR [\fB\-e \fIencoder\fR]] [\fB\-p\fR \fIpulse-width\fR]
[\fB\-y\fR]
.br
\fBbu0836\fR [\fB\-v\fR] [\fB\-d\fI device\fR] [\fB\-X\fR] [\fB\-O\fI file\fR] [\fB\-I\fI file\fR]
.br
\fBbu0836\fR [\fB\-h\fR|\fB\-\-version\fR]
'\"
'\"
'\"
'\"""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
.SH DESCRIPTION
.B bu0836
is an implementation of Leo Bodnar's configuration utilities \*(lqBU0836\ configuration.exe\*(rq and
\*(lqBU0836_encoders.exe\*(rq. All configuration changes are made in a local copy of the controller's
EEPROM and only copied to the device at the end after confirmation by the user, or whenever the
\fB\-\-sync\fR option is executed. The only exception is the \fB\-\-load\fR option, which flashes
the controller's EEPROM immediately. Before making any changes to the EEPROM its original state
should be saved to a file using the \fB\-\-save\fR option.
'\"
'\"
'\"
'\"""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
.SH OPTIONS
'\"
'\"
'\"
.SS "General Options"
'\"
'\"
'\"
.TP
.BR \-h ", " \-\-help
Show help screen and exit.
'\"""""
.TP
.B \-\-version
Show version number and exit.
'\"""""
.TP
.BR \-v ", " \-\-verbose
Increase verbosity level. The option can be used up to three times. It is evaluated before any
other option and can, thus, be placed anywhere.
'\"""""
.TP
.BR \-l ", " \-\-list
List BU0836 devices with \fIUSB bus id\fR, \fIvendor\fR, \fIproduct\fR, \fIserial number\fR,
and \fIfirmware version\fR. The output could look like in this example:
.LP
.RS 12
\fC$ bu0836 \-\-list
.br
\m[magenta]2:2\m[]     Leo Bodnar, BU0836A Interface, A1210\m[magenta]4\m[], v01.22
.br
\m[magenta]2:3\m[]     Leo Bodnar, BU0836A Interface, A121\m[magenta]16\m[], v01.22
.br
\m[magenta]2:9\m[]     Leo Bodnar, BU0836A Interface, A121\m[magenta]36\m[], v01.22\fR
.RE -12
.LP
.RS
Colorized elements can be used as search terms for the \fB\-\-device\fR option.
A trailing double angle marker (\m[green]<<\m[]) shows which device is currently selected (if any).
.RE
'\"""""
'\"
'\"
'\"
.SS "Device Options"
'\"
'\"
'\"
.TP
\fB\-d \fIdevice\fR, \fB\-\-device\fR=\fIdevice
Select device by \fIbus id\fR or \fIserial number\fR or an unambiguous ending of the
\fIserial number\fR. Using this option is only required if there is more than one
device attached. A single device is automatically selected. In the above example
the first device could be selected with \-\-device=2:2, \-\-device=A12104,
or just\ \-d4. Because two serial numbers end with 6, the second device would
have to be selected with at least\ \-d16.
'\"""""
.TP
.BR \-s ", " \-\-status
Show current device configuration. All default settings are shown in green color,
non-default values in red color. Axis indices are shown in bold font and with
preceding hash mark\ (\fC#\fR) if the axis is active.
'\"""""
.TP
.BR \-m ", " \-\-monitor
Continuously monitor a device's output until terminated with Ctrl-c.
'\"""""
.TP
.BR \-r ", " \-\-reset
Reset device configuration to \*(lqfactory default\*(rq. This is an equivalent of \-\-axes=0\-7
\-\-shut\-off=off \-\-invert=off \-\-zoom=off \-\-buttons=0\-31 \-\-encoder=off
\-\-pulse\-width=48ms.
'\"""""
.TP
.BR \-y ", " \-\-sync
Write all changes so far to the controller's EEPROM. If there are no uncommitted changes
left at exit, then the confirmation prompt is skipped, which makes this option useful for batch
processing.
'\"""""
.TP
.BR \-X ", " \-\-dump
Display the contents of the EEPROM image buffer. If used before any configuration
changes or right after a \fB\-\-sync\fR option it shows the actual EEPROM contents.
'\"""""
.TP
\fB\-O \fIfile\fR, \fB\-\-save\fR=\fIfile
Save EEPROM image buffer to \fIfile\fR. This should be done before making changes
to the device's EEPROM for the first time, so that its original state can be restored
if necessary.
'\"""""
.TP
\fB\-I \fIfile\fR, \fB\-\-load\fR=\fIfile
Load EEPROM image from \fIfile\fR and flash the controller's EEPROM with it. This
happens immediately and without asking for confirmation. Try to avoid using this
option! Applying the \fB\-\-reset\fR option instead should be enough.
'\"""""
'\"
'\"
'\"
.SS "Axis Options"
'\"
'\"
'\"
.TP
\fB\-a \fIlist\fR, \fB\-\-axes\fR=\fIlist
Select axes that the following axis operations shall be applied to. \fIlist\fR is an axis
number or number range, or a list thereof separated by commas, e\&.g\&. \*(lq0\*(rq or
\*(lq1,3,5\*(rq or \*(lq0\-5\*(rq or\ \*(lq0,2\-5,7\*(rq. Numbers must be in a range from
0 to\ 7. The list overrides a prior axis selection.
'\"""""
.TP
\fB\-i \fIinvert\fR, \fB\-\-invert\fR=\fIinvert
Set inverted mode for selected axes. The \fIinvert\fR string can be \*(lqon\*(rq and \*(lqoff\*(rq,
or 1 and\ 0.
'\"""""
.TP
\fB\-z \fIzoom\fR, \fB\-\-zoom\fR=\fIzoom
Set zoom mode for selected axes. The \fIzoom\fR string can be \*(lqon\*(rq and \*(lqoff\*(rq, or a
zoom factor in a range from 0 to\ 255. \*(lqon\*(rq corresponds to factor\ 198, and \*(lqoff\*(rq
to factor\ 0.
'\"""""
'\"
'\"
'\"
.SS "Button/Encoder Options"
'\"
'\"
'\"
.TP
\fB\-b \fIlist\fR, \fB\-\-buttons\fR=\fIlist
Select buttons that the following button operations shall be applied to. This works
just like the \fB\-\-axes\fR option, but with numbers ranging from 0 to\ 31.
'\"""""
.TP
\fB\-e \fIencoder\fR, \fB\-\-encoder\fR=\fIencoder
Set encoder mode for selected buttons (and their associated pair-buttons). \fIencoder\fR can be
.LP
.RS 12
\*(lqoff\*(rq or 0 for normal button function,
.br
\*(lq1:1\*(rq or 1 for quarter wave encoder,
.br
\*(lq1:2\*(rq or 2 for half wave encoder,
.br
\*(lq1:4\*(rq or 3 for full wave encoder (one detent produces four state changes).
.RE -12
.LP
.RS
A button pair always consists of an even numbered button and the next odd numbered one.
It doesn't matter which of them is used for configuring an encoder. \-b0, \-b1, and\ \-b0,1
yield the same result. Most encoders are of the 1:2\ kind, so this is a good mode to start
with if one doesn't know the details.
.RE
'\"""""
.TP
\fB\-p \fIpulse-width\fR, \fB\-\-pulse\-width\fR=\fIpulse-width
Set pulse width for all encoders. \fIpulse-width\fR is either an integer number in the range
1 to\ 11, which, multiplied with\ 8, denotes the pulse width in \fIms\fR. Alternatively, the pulse
width can be entered directly in \fIms\fR by appending \*(lqms\*(rq. This number is then rounded to
the nearest available 8\ ms step. \-p48ms is equivalent to\ \-p6, and is the factory default.
'\"""""
'\"
'\"
'\"
'\"""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
.SH EXAMPLES
.TP
\fC$ bu0836 \-\-reset \-\-sync
Reset controller to \fI\*(lqfactory default\*(rq\fR and commit the changes immediately.
'\"""""
.TP
\fC$ bu0836 \-\-axes=0\-7 \-\-invert=on
Set inverted mode for all axes. The user will be asked for confirmation before committing.
'\"""""
.TP
\fC$ bu0836 \-\-buttons=0,4,9 \-\-encoder=1:2 \-\-sync
Configure button pairs 0/1, 4/5, and\ 8/9 for half wave encoder and commit the changes immediately.
'\"""""
.TP
\fC$ bu0836 \-r \-a0\-7 \-i1 \-b0,4,9 \-e2 \-y \-s
The first three examples combined using short options and followed by a status report.
'\"
'\"
'\"
'\"""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
.SH NOTES
Theoretically, \fBbu0836\fR supports 31 different devices \(em the same that Leo Bodnar's tools support. Only few
of those devices offer axis configuration \fIand\fR encoder configuration \(em most support only
encoders. So, if the \fB\-\-status\fR option doesn't display axes, this probably means that your
device doesn't support them.
.P
\fBbu0836\fR enumerates axes and buttons beginning with\ 0, because this is more logical in a
programming context and also the way the Linux kernel sees it. Bodnar's tools, however, start
numbering with 1, and so does the firmware.
.P
In case of problems don't hesitate to contact the author.
'\"
'\"
'\"
'\"""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
.SH AUTHOR
2010 \- Melchior FRANZ <\m[blue]melchior.franz@gmail.com\m[]>


bu0836.cxx
// bu0836 device manager
//
// Copyright (C) 2010  Melchior FRANZ  <melchior.franz@gmail.com>
//
// This program is free software; you can redistribute it and/or
// modify it under the terms of the GNU General Public License as
// published by the Free Software Foundation; either version 2 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful, but
// WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
// General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
//
#include <cstring> // memcpy
#include <fstream>
#include <iomanip>
#include <iostream>
#include <signal.h>
#include <sstream>

#include "bu0836.hxx"
#include "hid.hxx"
#include "logging.hxx"
#include "options.h"

using namespace std;
using namespace logging;



namespace bu0836 {

namespace {

#ifdef VALGRIND
bool interrupted = true;
#else
bool interrupted = false;
#endif



void interrupt_handler(int)
{
	log(BULK) << "Interrupted" << endl;
	interrupted = true;
}



const char *usb_strerror(int errno)
{
	switch (errno) {
	case LIBUSB_SUCCESS:
		return "success";
	case LIBUSB_ERROR_IO:
		return "input/output error";
	case LIBUSB_ERROR_INVALID_PARAM:
		return "invalid parameter";
	case LIBUSB_ERROR_ACCESS:
		return "access denied (insufficient permissions)";
	case LIBUSB_ERROR_NO_DEVICE:
		return "no such device (it may have been disconnected)";
	case LIBUSB_ERROR_NOT_FOUND:
		return "entity not found";
	case LIBUSB_ERROR_BUSY:
		return "resource busy";
	case LIBUSB_ERROR_TIMEOUT:
		return "operation timed out";
	case LIBUSB_ERROR_OVERFLOW:
		return "overflow";
	case LIBUSB_ERROR_PIPE:
		return "pipe error";
	case LIBUSB_ERROR_INTERRUPTED:
		return "system call interrupted (perhaps due to signal)";
	case LIBUSB_ERROR_NO_MEM:
		return "insufficient memory";
	case LIBUSB_ERROR_NOT_SUPPORTED:
		return "operation not supported or unimplemented on this platform";
	case LIBUSB_ERROR_OTHER:
		return "other error";
	default:
		return "unknown error code";
	}
}



string bcd2str(int n)
{
	ostringstream o;
	o << hex << ((n >> 12) & 0xf) << ((n >> 8) & 0xf) << '.' << ((n >> 4) & 0xf) << (n & 0xf);
	return o.str();
}



string strip(const string &s)
{
	const char spaces[] = " \t\n\r\b";
	string::size_type start = s.find_first_not_of(spaces, 0);
	string::size_type end = s.find_last_not_of(spaces, string::npos);
	return start == string::npos || end == string::npos ? "" : s.substr(start, end - start + 1);
}

} // namespace



controller::controller(libusb_device_handle *handle, libusb_device *device, libusb_device_descriptor desc,
		int capabilities) :
	_handle(handle),
	_device(device),
	_desc(desc),
	_active_axes(0),
	_capabilities(capabilities),
	_hid_descriptor(0),
	_claimed(false),
	_kernel_detached(false),
	_dirty(false)
{
	ostringstream s;
	s << int(libusb_get_bus_number(_device)) << ':' << int(libusb_get_device_address(_device));
	_bus_address = s.str();

	s.str("");
	s << hex << setfill('0') << setw(4) << _desc.idVendor << ':' << setw(4) << _desc.idProduct;
	_id = s.str();

	_release = bcd2str(_desc.bcdDevice);

	unsigned char buf[256];
	if (_desc.iManufacturer && libusb_get_string_descriptor_ascii(_handle, _desc.iManufacturer, buf, sizeof(buf)) > 0)
		_manufacturer = strip(string((char *)buf));

	if (_desc.iProduct && libusb_get_string_descriptor_ascii(_handle, _desc.iProduct, buf, sizeof(buf)) > 0)
		_product = strip(string((char *)buf));

	if (_desc.iSerialNumber && libusb_get_string_descriptor_ascii(_handle, _desc.iSerialNumber, buf, sizeof(buf)) > 0)
		_serial = strip(string((char *)buf));

	_jsid = _manufacturer;
	if (!_jsid.empty() && !_product.empty())
		_jsid += ' ';
	_jsid += _product;
	if (!_jsid.empty() && !_serial.empty())
		_jsid += ' ';
	_jsid += _serial;
}



controller::~controller()
{
	int ret;
	if (_claimed) {
		ret = libusb_release_interface(_handle, _INTERFACE);
		if (ret < 0)
			log(ALERT) << "libusb_release_interface: " << usb_strerror(ret) << endl;
	}

	if (_kernel_detached) {
		ret = libusb_attach_kernel_driver(_handle, _INTERFACE);
		if (ret < 0)
			log(ALERT) << "libusb_attach_kernel_driver: " << usb_strerror(ret) << endl;
	}

	libusb_close(_handle);
	delete [] _hid_descriptor;
}



int controller::claim()
{
	int ret;
	if (!_kernel_detached && libusb_kernel_driver_active(_handle, _INTERFACE)) {
		ret = libusb_detach_kernel_driver(_handle, _INTERFACE);
		if (ret < 0) {
			log(ALERT) << "libusb_detach_kernel_driver: " << usb_strerror(ret) << endl;
			return ret;
		}
		_kernel_detached = true;
	}

	if (!_claimed) {
		ret = libusb_claim_interface(_handle, _INTERFACE);
		if (ret < 0) {
			log(ALERT) << "libusb_claim_interface: " << usb_strerror(ret) << endl;
			return ret;
		}
		_claimed = true;
		ret = parse_hid();
		if (ret)
			return ret;

		_active_axes = get_active_axes(_hid.data()[0]);

		ret = get_eeprom();
		if (ret)
			return ret;
	}

	return 0;
}



int controller::parse_hid()
{
	unsigned char buf[255];
	int ret = libusb_get_descriptor(_handle, LIBUSB_DT_HID, 0, buf, sizeof(buf));
	if (ret < 0) {
		log(ALERT) << "libusb_get_descriptor: " << usb_strerror(ret) << endl;
		return ret;
	}

	_hid_descriptor = reinterpret_cast<usb_hid_descriptor *>(new unsigned char[ret]);
	memcpy(_hid_descriptor, buf, ret);
	int numreports = _hid_descriptor->bNumDescriptors;

	log(BULK) << "HID: type=" << int(_hid_descriptor->bDescriptorType) << "  num=" << numreports << endl;

	for (int i = 0; i < numreports; i++) {
		int len = _hid_descriptor->wDescriptorLength(i);
		log(BULK) << "REPORT " << i << ": type=" << int(_hid_descriptor->descriptors[i].bDescriptorType)
				<< "  len=" << len << endl;

		unsigned char *buf = new unsigned char[len];
		ret = libusb_get_descriptor(_handle, LIBUSB_DT_REPORT, 0, buf, len);
		if (ret < 0)
			log(ALERT) << "libusb_get_descriptor/LIBUSB_DT_REPORT: " << usb_strerror(ret) << endl;
		else if (ret != len)
			log(ALERT) << "libusb_get_descriptor/LIBUSB_DT_REPORT: only " << ret << " of " << len
					<< " bytes delivered" << endl;
		else
			_hid.parse(buf, ret);

		delete [] buf;
	}
	return 0;
}



int controller::get_eeprom()
{
	unsigned char buf[17];
	int progress = 0xffff;
	int maxtries = 50;
	while (progress && maxtries--) {
		int ret = libusb_control_transfer(_handle, /* CLASS SPECIFIC REQUEST IN */ 0xa1,
				/* GET_REPORT */ 0x01, /* FEATURE */ 0x0300, 0, buf, sizeof(buf), 1000 /* ms */);
		if (ret < 0) {
			log(ALERT) << "get_eeprom/libusb_control_transfer: " << usb_strerror(ret) << endl;
			return -1;
		}
		if (ret != sizeof(buf))
			continue;
		if (buf[0] & 0x0f)
			continue;
		progress &= ~(1 << (buf[0] >> 4));
		memcpy(reinterpret_cast<uint8_t *>(&_eeprom) + buf[0], buf + 1, 16);
	}
	if (!maxtries) {
		log(ALERT) << "get_eeprom: unable to read whole EEPROM" << endl;
		return -2;
	}
	return 0;
}



int controller::set_eeprom(unsigned int from, unsigned int to)
{
	if (to < from || to >= sizeof(_eeprom))
		throw(ORIGIN"set_eeprom: internal error");

	unsigned char buf[2];
	unsigned char *eeprom = reinterpret_cast<uint8_t *>(&_eeprom);
	for (unsigned int i = from; i <= to; i++) {
		buf[0] = i;
		buf[1] = eeprom[i];
		int ret = libusb_control_transfer(_handle, /* CLASS SPECIFIC REQUEST OUT */ 0x21,
				/* SET_REPORT */ 0x09, /* FEATURE */ 0x0300, 0, buf, 2, 1000 /* ms */);
		if (ret < 0) {
			log(ALERT) << "set_eeprom/libusb_control_transfer: " << usb_strerror(ret) << endl;
			return -1;
		}
	}
	return 0;
}



int controller::save_image_file(const char *path)
{
	ofstream file(path, ofstream::binary | ofstream::trunc);
	if (!file)
		throw string("cannot write to '") + path + '\'';
	file.write(reinterpret_cast<const char *>(&_eeprom), sizeof(_eeprom));

	file.seekp(0, ofstream::end);
	if (file.tellp() != sizeof(_eeprom))
		throw string("file '") + path + "' has wrong size";

	file.close();
	return 0;
}



int controller::load_image_file(const char *path)
{
	ifstream file(path, ifstream::binary);
	if (!file)
		throw string("cannot read from '") + path + '\'';
	file.read(reinterpret_cast<char *>(&_eeprom), sizeof(_eeprom));

	file.seekg(0, ifstream::end);
	if (file.tellg() != sizeof(_eeprom))
		throw string("file '") + path + "' has wrong size";

	file.close();
	return 0;
}



int controller::show_input_reports()
{
	if (_hid.data().empty()) {
		log(ALERT) << "show_input_reports: no hid data" << endl;
		return 1;
	}

	struct sigaction sa;
	sigemptyset(&sa.sa_mask);
	sa.sa_flags = 0;
	sa.sa_handler = interrupt_handler;
	sigaction(SIGHUP, &sa, NULL);
	sigaction(SIGINT, &sa, NULL);
	sigaction(SIGTERM, &sa, NULL);
	sigaction(SIGQUIT, &sa, NULL);

	unsigned char buf[1024];
	int len;
	do {
		int ret = libusb_interrupt_transfer(_handle, LIBUSB_ENDPOINT_IN | 1, buf, sizeof(buf), &len, 100 /* ms */);
		if (ret < 0) {
			log(ALERT) << "show_input_reports/libusb_interrupt_transfer: "
					<< usb_strerror(ret) << ", " << len << endl;
			sleep(2);
			continue;
		}

		log(BULK) << endl << bytes(buf, len) << endl;
		print_input(_hid.data()[0], buf);
		cout << endl;

		const struct timespec ts = {0, 100000};
		nanosleep(&ts, 0);
	} while (!interrupted);
	return 0;
}



void controller::print_input(hid::hid_main_item *item, const unsigned char *data)
{
	vector<hid::hid_main_item *>::const_iterator it, end = item->children().end();
	for (it = item->children().begin(); it != end; ++it)
		print_input(*it, data);

	if (item->type() != hid::INPUT || (item->data_type() & 1)) // no padding
		return;

	uint32_t colltype = item->parent() ? item->parent()->data_type() : 0;

	vector<hid::hid_value>::const_iterator vit, vend = item->values().end();
	int index = 0;
	for (vit = item->values().begin(); vit != vend; ++vit, index++) {
		uint32_t v = vit->get_unsigned(data);
		if (colltype == 0) { // axis (physical)
			int num = vit->usage() - 48; // Usage 'X' == 0x30
			double norm = double(v) / item->global().logical_maximum;
			cout << "A" << num << '=' << cyan << setfill(' ') << setw(4) << v << reset
					<< " (" << magenta << fixed << setprecision(4) << norm << reset << ") ";

		} else if (item->global().usage_table == 0x09) { // button
			if (index == 16)
				cout << endl;
			cout << "B" << setw(2) << setfill('0') << index << '='
					<< (v ? red : green) << v << reset << ' ';

		} else if (item->global().usage_table == 0x01 && vit->usage() == 0x39) { // hat
			cout << "H" << '=' <<  brown << v << reset << ' ';

		} else {
			log(WARN) << "something " << vit->name() << " " << vit->usage() << endl;
		}
	}
	cout << endl;
}



int controller::get_active_axes(hid::hid_main_item *item)
{
	int axes = 0;
	vector<hid::hid_main_item *>::const_iterator it, end = item->children().end();
	for (it = item->children().begin(); it != end; ++it)
		axes |= get_active_axes(*it);

	if (item->type() == hid::INPUT && !(item->data_type() & 1) // non-padding input
			&& item->parent() && item->parent()->data_type() == 0) { // collection type

		vector<hid::hid_value>::const_iterator vit, vend = item->values().end();
		for (vit = item->values().begin(); vit != vend; ++vit)
			axes |= 1 << (vit->usage() - 48);
	}
	return axes;
}



void controller::set_encoder_mode(int b, int mode)
{
	uint16_t b0 = _eeprom.rotenc0[0] | (_eeprom.rotenc0[1] << 8);
	uint16_t b1 = _eeprom.rotenc1[0] | (_eeprom.rotenc1[1] << 8);
	uint16_t mask = 1 << b / 2;
	b0 = mode & 1 ? b0 | mask : b0 & ~mask;
	b1 = mode & 2 ? b1 | mask : b1 & ~mask;
	_eeprom.rotenc0[0] = b0 & 0xff, _eeprom.rotenc0[1] = (b0 >> 8) & 0xff;
	_eeprom.rotenc1[0] = b1 & 0xff, _eeprom.rotenc1[1] = (b1 >> 8) & 0xff;
	_dirty = true;
}



int controller::get_encoder_mode(int b) const
{
	b /= 2;
	uint16_t b0 = ((_eeprom.rotenc0[0] | (_eeprom.rotenc0[1] << 8)) >> b) & 1;
	uint16_t b1 = ((_eeprom.rotenc1[0] | (_eeprom.rotenc1[1] << 8)) >> b) & 1;
	return b0 | (b1 << 1);
}



manager::manager(int debug_level)
{
	int ret = libusb_init(_CONTEXT);
	if (ret < 0)
		throw string("libusb_init: ") + usb_strerror(ret);
	libusb_set_debug(_CONTEXT, debug_level);

	libusb_device **list;
	int num = libusb_get_device_list(_CONTEXT, &list);
	if (num < 0)
		throw string("libusb_get_device_list: ") + usb_strerror(ret);

	for (int i = 0; i < num; i++) {
		libusb_device_handle *handle;
		ret = libusb_open(list[i], &handle);
		if (ret) {
			log(ALERT) << "error: libusb_open: " << usb_strerror(ret) << endl;
			continue;
		}

		libusb_device *dev = libusb_get_device(handle);
		libusb_device_descriptor desc;
		ret = libusb_get_device_descriptor(dev, &desc);

		int capabilities = 0;

		if (ret) {
			log(ALERT) << "error: libusb_get_device_descriptor: " << usb_strerror(ret) << endl;

		} else if (desc.idVendor == 0x16c0) { // VOTI
			switch (desc.idProduct) {
			case 0x05b5: // BU0836
				capabilities = INVERT | ENCODER1 | ENCODER2;
				break;
			case 0x278a:
			case 0x2795:
			case 0x05ba: // BU0836A
				capabilities = INVERT | ZOOM | ENCODER1 | ENCODER2;
				break;
			case 0x05b7: case 0x27bb: case 0x27be: case 0x27c4: case 0x27b9:
			case 0x27bd: case 0x279a: case 0x27a8: case 0x27a3: case 0x05bb:
			case 0x279b: case 0x27c7: case 0x27c8: case 0x27c9: case 0x27ca:
			case 0x27cc: case 0x27cd: case 0x27cf: case 0x27d0: case 0x27d1:
			case 0x27d2: case 0x2794:
				capabilities = ENCODER1 | ENCODER2;
				break;
			}

		} else if (desc.idVendor == 0x1dd2) { // Leo Bodnar
			switch (desc.idProduct) {
			case 0x1001: // BU0836X
			case 0x1002: case 0x2001: case 0x2002: case 0x2003:
				capabilities = ENCODER1 | ENCODER2;
				break;
			case 0x200a:
				capabilities = INVERT | ZOOM | ENCODER1 | ENCODER2;
				break;
			}
		}

		if (!capabilities) {
			libusb_close(handle);
			continue;
		}

		if (desc.bcdDevice < 0x0118)
			capabilities &= ~ZOOM;
		if (desc.bcdDevice < 0x0120)
			capabilities &= ~ENCODER1;
		if (desc.bcdDevice < 0x0121)
			capabilities &= ~ENCODER2;

		_devices.push_back(new controller(handle, dev, desc, capabilities));
	}
	libusb_free_device_list(list, 1);
	_selected = size() == 1 ? _devices[0] : 0;
}



manager::~manager()
{
	vector<controller *>::const_iterator it, end = _devices.end();
	for (it = _devices.begin(); it != end; ++it)
		delete *it;
	libusb_exit(_CONTEXT);
}



int manager::select(const string &which)
{
	int num = 0;
	vector<controller *>::const_iterator it, end = _devices.end();
	for (it = _devices.begin(); it != end; ++it)
		if ((*it)->serial().rfind(which) == (*it)->serial().size() - which.size()
				|| (*it)->bus_address() == which)
			_selected = *it, num++;

	if (num != 1)
		_selected = 0;
	return num;
}

} // namespace bu0836
